### PR TITLE
feat: bring back WebGPU

### LINKED ISSUE #21514: Reintroduce WebGPU API
The WebGPU API was removed "temporarily" for 1.32.0 in https://github.com/denoland/deno/pull/18094. We are stuck on version 1.31.3 of Deno until then.

What are your plans to reintroduce this feature? Is there anything that blocks you from adding WebGPU support again? It is an amazing feature and it enables a wealth of new applications that no other runtime offers.

### PR DESCRIPTION
**Pull request by crowlKats**
Fixes #21514

### CODE REVIEW

**Comment by sigmaSd** [comment]
does this not impact startup time anymore ?

**Comment by hastebrot** [comment]
if it impacts startup time then it should be put behind a flag, if this helps. otherwise everything will slowdown including `deployctl` (deno deploy cli) which currently is 20 times faster then `az` (azure cli tool) just to output the help text.

File: .cargo/config.toml [hunk 1]
```diff
@@ -11,8 +11,14 @@ rustflags = [
...
 ]
 
+[target.x86_64-apple-darwin]
+rustflags = ["-C", "link-args=-weak_framework Metal -weak_framework MetalPerformanceShaders -weak_framework QuartzCore"]
+
 [target.aarch64-apple-darwin]
-rustflags = ["-C", "link-arg=-fuse-ld=lld"]
+rustflags = [
+  "-C",
+  "link-args=-fuse-ld=lld -weak_framework Metal -weak_framework MetalPerformanceShaders -weak_framework QuartzCore",
```
Comment (littledivy) [inline .cargo/config.toml:20]: `CoreGraphics` too?

**Comment by crowlKats** [comment]
We are working on reducing the startup time. A flag would not help. Also, the startup time originally was increased by 3ms, and now is at 1ms increase, and we are working on further reducing it.

**Comment by chances** [comment]
> Requires a wgpu release...

@crowlKats What is missing from wgpu, exactly?

**Comment by crowlKats** [comment]
> > Requires a wgpu release...
> 
> 
> 
> @crowlKats What is missing from wgpu, exactly?

@chances due to our setup with the wgpu repo, we require the js & rust code to be pulled from a commit that is a release of wgpu.
The blockers for wgpu release from my end is a core-graphics-types release and then a metal-rs release.

File: ext/webgpu/01_webgpu.js [hunk 2]
```diff
@@ -0,0 +1,6830 @@
...
+    webidl.illegalConstructor();
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:2461]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUShaderModulePrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 3]
```diff
@@ -0,0 +1,6830 @@
...
+}
+GPUObjectBaseMixin("GPUShaderModule", GPUShaderModule);
+
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:2464]:
```suggestion
const GPUShaderModulePrototype = GPUShaderModule.prototype;
```

File: ext/webgpu/01_webgpu.js [hunk 4]
```diff
@@ -0,0 +1,6830 @@
...
+    return bindGroupLayout;
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:2550]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUComputePipelinePrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 5]
```diff
@@ -0,0 +1,6830 @@
...
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:2621]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPURenderPipelinePrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 6]
```diff
@@ -0,0 +1,6830 @@
...
+    return commandBuffer;
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:3316]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUCommandEncoderPrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 7]
```diff
@@ -0,0 +1,6830 @@
...
+    );
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:3918]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPURenderPassEncoderPrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 8]
```diff
@@ -0,0 +1,6830 @@
...
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4188]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUComputePassEncoderPrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 9]
```diff
@@ -0,0 +1,6830 @@
...
+    webidl.illegalConstructor();
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4233]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUCommandBufferPrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 10]
```diff
@@ -0,0 +1,6830 @@
...
+}
+GPUObjectBaseMixin("GPUCommandBuffer", GPUCommandBuffer);
+
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4236]: const GPUCommandBufferPrototype = GPUCommandBuffer.prototype;

File: ext/webgpu/01_webgpu.js [hunk 11]
```diff
@@ -0,0 +1,6830 @@
...
+    );
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4617]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPURenderBundleEncoderPrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 12]
```diff
@@ -0,0 +1,6830 @@
...
+    webidl.illegalConstructor();
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4662]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPURenderBundlePrototype, this),
        keys: [
          "label",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/01_webgpu.js [hunk 13]
```diff
@@ -0,0 +1,6830 @@
...
+}
+GPUObjectBaseMixin("GPURenderBundle", GPURenderBundle);
+
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4665]:
```suggestion
const GPURenderBundlePrototype = GPURenderBundle.prototype;
```

File: ext/webgpu/01_webgpu.js [hunk 14]
```diff
@@ -0,0 +1,6830 @@
...
+    this[_count]();
+  }
+
+  [SymbolFor("Deno.privateCustomInspect")](inspect) {
+    return `${this.constructor.name} ${
+      inspect({
+        label: this.label,
+      })
+    }`;
+  }
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:4728]:
```suggestion
  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUQuerySetPrototype, this),
        keys: [
          "label",
          "type",
          "count",
        ],
      }),
      inspectOptions,
    );
  }
```

File: ext/webgpu/02_surface.js [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+    const prefix = "Failed to execute 'present' on 'GPUCanvasContext'";
+    const device = assertDevice(this[_currentTexture], {
+      prefix,
+      context: "this",
+    });
+    ops.op_webgpu_surface_present(device.rid, this[_surfaceRid]);
+    this[_currentTexture].destroy();
+    this[_currentTexture] = undefined;
+  }
+}
```
Comment (petamoriken) [inline ext/webgpu/02_surface.js:134]:
```suggestion

  [SymbolFor("Deno.privateCustomInspect")](inspect, inspectOptions) {
    return inspect(
      createFilteredInspectProxy({
        object: this,
        evaluate: ObjectPrototypeIsPrototypeOf(GPUCanvasContextPrototype, this),
        keys: [
          "canvas",
        ],
      }),
      inspectOptions,
    );
  }
}
```

**Review by petamoriken — COMMENTED** [review]
Added many suggestions related to inspect 🙇

File: ext/webgpu/01_webgpu.js [hunk 16]
```diff
@@ -0,0 +1,7044 @@
...
+  /** @returns {GPUSupportedLimits} */
+  get limits() {
+    webidl.assertBranded(this, GPUAdapterPrototype);
+    return this[_adapter].limits;
+  }
+  /** @returns {boolean} */
+  get isFallbackAdapter() {
+    webidl.assertBranded(this, GPUAdapterPrototype);
+    webidl.assertBranded(this, GPUAdapterPrototype);
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:309]: There is a duplicate 🙇

**Comment by crowlKats** [comment]
@petamoriken PTAL at primordials in CI

File: ext/webgpu/01_webgpu.js [hunk 17]
```diff
@@ -0,0 +1,7055 @@
...
+      selfContext: "this",
+      resourceContext: "Argument 1",
+    });
+    const { err } = ops.op_webgpu_write_buffer(
+      device.rid,
+      bufferRid,
+      bufferOffset,
+      dataOffset,
+      size,
+      new Uint8Array(ArrayBufferIsView(data) ? data.buffer : data),
```
Comment (petamoriken) [inline ext/webgpu/01_webgpu.js:1691]: `TypedArrayPrototypeGetBuffer` should be used for `TypedArray`, and `DataViewPrototypeGetBuffer` should be used for `DataView`.

https://github.com/denoland/deno/blob/3f96e5a29a88afafcef0f17458b2800b2db316ee/ext/web/08_text_encoding.js#L112-L124

**Review by petamoriken — COMMENTED** [review]
@crowlKats prefer-primordials lint worked fine in my local environment. so I wrote down the result :bow:

**Comment by KnorpelSenf** [comment]
This fixes #21514, feel free to update the description

**Comment by ry** [comment]
@crowlKats can you provide start up time benchmarks? Also max RSS for a blank file?

**Comment by crowlKats** [comment]
@ry 
Screenshot 2023-12-08 at 16 41 42

**Comment by crowlKats** [comment]
Memory:
Screenshot 2023-12-08 at 16 59 10
Screenshot 2023-12-08 at 16 59 47

File: cli/js/40_testing.js [hunk 18]
```diff
@@ -142,7 +142,10 @@ const OP_DETAILS = {
     "op_utime_async": ["change file timestamps", "awaiting the result of a `Deno.utime` call"],
     "op_host_recv_message": ["receive a message from a web worker", "terminating a `Worker`"],
     "op_host_recv_ctrl": ["receive a message from a web worker", "terminating a `Worker`"],
-    "op_ws_close": ["close a WebSocket", "awaiting until the `close` event is emitted on a `WebSocket`, or the `WebSocketStream#closed` promise resolves"],
+    "op_webgpu_buffer_get_map_async": ["map a WebGPU buffer", "awaiting the result of a `GPUBuffer#mapAsync` call"],
+	  "op_webgpu_request_adapter": ["request a WebGPU adapter", "awaiting the result of a `navigator.gpu.requestAdapter` call"],
+	  "op_webgpu_request_device": ["request a WebGPU device", "awaiting the result of a `GPUAdapter#requestDevice` call"],
+	  "op_ws_close": ["close a WebSocket", "awaiting until the `close` event is emitted on a `WebSocket`, or the `WebSocketStream#closed` promise resolves"],
```
Comment (bartlomieju) [inline cli/js/40_testing.js:148]: Ouch, formatting seems off here. @dsherret any ideas?

File: cli/tests/unit/webgpu_test.ts [hunk 19]
```diff
@@ -0,0 +1,242 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import { assert, assertEquals } from "./test_util.ts";
+
+let isCI: boolean;
+try {
+  isCI = (Deno.env.get("CI")?.length ?? 0) > 0;
```
Comment (bartlomieju) [inline cli/tests/unit/webgpu_test.ts:7]: This is 4th place we define this env var :) we should move it to `test_util.ts` in a follow up. Can you handle that @crowlKats?

File: cli/tests/unit/webgpu_test.ts [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+// Skip these tests on linux CI, because the vulkan emulator is not good enough
+// yet, and skip on macOS CI because these do not have virtual GPUs.
+const isLinuxOrMacCI =
+  (Deno.build.os === "linux" || Deno.build.os === "darwin") && isCI;
+// Skip these tests in WSL because it doesn't have good GPU support.
+const isWsl = await checkIsWsl();
+
+Deno.test({
+  permissions: { read: true, env: true },
+  ignore: isWsl || isLinuxOrMacCI,
```
Comment (bartlomieju) [inline cli/tests/unit/webgpu_test.ts:21]: So are any of these tests run on our CI after all?

File: ext/node/polyfill.rs [hunk 21]
```diff
@@ -19,6 +19,10 @@ macro_rules! generate_builtin_node_module_lists {
       $(
         concat!("node:", $module_name),
       )+
+    "ext:deno_webidl/00_webidl.js",
+    "ext:deno_web/02_event.js",
+    "ext:deno_web/01_dom_exception.js",
+    "ext:deno_console/01_console.js",
```
Comment (bartlomieju) [inline ext/node/polyfill.rs:25]: Revert, unrelated change

File: ext/webgpu/02_surface.js [hunk 22]
```diff
@@ -0,0 +1,236 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+// @ts-check
+/// <reference path="../../core/lib.deno_core.d.ts" />
+/// <reference path="../web/internal.d.ts" />
+/// <reference path="../web/lib.deno_web.d.ts" />
+/// <reference path="./lib.deno_webgpu.d.ts" />
+
+const core = globalThis.Deno.core;
```
Comment (bartlomieju) [inline ext/webgpu/02_surface.js:9]: Is this file executed? If not can we remove it? If it is then please rework to use `ext:core/mod.js` import.

File: ext/webgpu/README.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
+# deno_webgpu
+
+This op crate implements the WebGPU API as defined in
+https://gpuweb.github.io/gpuweb/ in Deno. The implementation targets the spec
+draft as of February 22, 2021. The spec is still very much in flux. This op
+crate tries to stay up to date with the spec, but is constrained by the features
+implemented in our GPU backend library [wgpu](https://github.com/gfx-rs/wgpu).
```
Comment (bartlomieju) [inline ext/webgpu/README.md:7]: Is this up to date?

File: runtime/js/98_global_scope.js [hunk 24]
```diff
@@ -190,7 +229,49 @@ const numCpus = memoizeLazy(() => ops.op_bootstrap_numcpus());
...
 ObjectDefineProperties(Navigator.prototype, {
+  gpu: {
+    configurable: true,
+    enumerable: true,
+    get() {
+      webidl.assertBranded(this, NavigatorPrototype);
+      loadWebGPU();
+      return webgpu.gpu;
+    },
+  },
```
Comment (bartlomieju) [inline runtime/js/98_global_scope.js:274]: What about `set()`? I think on the web one can override this value

File: cli/js/40_testing.js (hunk 18 above)
Comment (dsherret) [reply cli/js/40_testing.js:148]: See line 80.

File: runtime/js/98_global_scope.js (hunk 24 above)
Comment (crowlKats) [reply runtime/js/98_global_scope.js:274]: nope, not overwritable, however all the navigator getters should have no-op setters. I'll fix that up in a follow up with tests.

File: ext/webgpu/02_surface.js (hunk 22 above)
Comment (crowlKats) [reply ext/webgpu/02_surface.js:9]: I'll update the import, however it is not executed. this is used for embedding cases only

File: cli/tests/unit/webgpu_test.ts (hunk 19 above)
Comment (crowlKats) [reply cli/tests/unit/webgpu_test.ts:7]: sure

File: cli/tests/unit/webgpu_test.ts (hunk 20 above)
Comment (crowlKats) [reply cli/tests/unit/webgpu_test.ts:21]: windows ci

File: ext/webgpu/README.md (hunk 23 above)
Comment (crowlKats) [reply ext/webgpu/README.md:7]: updated the date

**Review by bartlomieju — APPROVED** [review]

**Comment by lino-levan** [comment]
Nice work Leo (and the whole Deno team). Excited to see this brought back into Deno.
